### ISSUE TITLE
Optional chaining does not count as branch for coverage

### ISSUE THREAD

**Issue by chrismiceli** [istanbuljs/istanbuljs#516]
Testing code that is using optional chaining @babel/plugin-proposal-optional-chaining as outlined here https://babeljs.io/docs/en/next/babel-plugin-proposal-optional-chaining.html does not count usages as branches for test coverage.
To Reproduce

Create a simple unit of code to test:

```
export const cGetter = (a) => {
  let first = a?.c;
  return first;
}
```

And create a test that only tests the defined case:
```
import {cGetter} from './getter';
test('gets c', () => {
  expect(cGetter({c: 'hi'})).toEqual('hi');
});
```
Expected behavior

I would expect 50% branch coverage since the case wehre a is falsy is unchecked.
Link to repl or repo (highly encouraged)

https://github.com/chrismiceli/jest-optional-chaining
envinfo
```
  System:
    OS: Windows 10 10.0.18363
    CPU: (8) x64 Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
  Binaries:
    Node: 12.14.0 - C:\Program Files\nodejs\node.EXE
    npm: 6.13.4 - C:\Program Files\nodejs\npm.CMD
  npmPackages:
    jest: ^24.9.0 => 24.9.0
```

**Comment by coreyfarrell** [comment]
This is actually an issue with istanbul-lib-instrument so I've transferred the issue to our monorepo.

**Comment by rijkvanzanten** [comment]
Is there a workaround available to ignore the optional chaining in the coverage reports?

Like the problem mentioned above, I'm getting "Uncovered lines" mentioned for:

```js
const status = error.response?.status;
const code = error.response?.data?.error?.code;
```

but not for

```js
const status = error.response && error.response.status;
const code = error.response && error.response.data && error.response.data.error && error.response.data.error.code;
```

Adding a bunch of `/* istanbul ignore next */` comments works, but doesn't feel like the right approach:

```js
/* istanbul ignore next */
const status = error.response?.status;
/* istanbul ignore next */
const code = error.response?.data?.error?.code;
```

**Comment by coreyfarrell** [comment]
```js
// NOT the same as optional chaining
const status1 = error.response && error.response.status;

// Optional chaining
const status2 = error.response?.status;

// Transpiled optional chaining (approximate)
let response_;
const status3 = (response_ = error.response, response_ !== null && response_ !== void 0 ? response._status : void 0);
```

Your coverage is based on the generated code which has code that you cannot see.  istanbul can only work with what it is given.  There is currently no solution or work-around for this issue.

**Comment by fabiomirandaa** [comment]
Does the problem continue?

**Comment by crestenstclair** [comment]
This issue continues to happen, I'm experiencing it in my code base

**Comment by chrismiceli** [comment]
Yes, the problem persists as exhibited by the linked repo https://github.com/chrismiceli/jest-optional-chaining/tree/master.

**Comment by joan-saum** [comment]
There is no plan about working on this issue ?

**Comment by anmolsinha1793** [comment]
Is there an update on the same?

**Comment by blephy** [comment]
Is there any work planned for this issue ?

**Comment by JavierBrooktec** [comment]
this worked for me if https://stackoverflow.com/a/66949633

**Comment by PRossetti** [comment]
I'm having the same issue here. Code like this would be mark as 100% covered by jest

if (body && body.foo && body.foo.faa) { ... }

But this equivalent code won't. The ?. won't be marked as covered.
if (body?.foo?.faa) { ... }

Having an easy elegant way of getting pass by this behaviour would be very much appreciated.

**Comment by balboFK** [comment]
This is a commom problem for me aswell.

**Comment by petarzarkov** [comment]
Any update on this?

**Comment by GustavoSchuler** [comment]
There is any update on this issue?

**Comment by sheilf-synapse** [comment]
We started to replace optional chains with lodash `get`

**Comment by chrismiceli** [comment]
@sheilf-synapse, that still hide branches that code coverage cannot indicate if covered or not.

**Comment by japamatSynapse** [comment]
for us, it's a workaround that gets our branch coverage up. Previously jest would not consider optionally chained values as covered in branches, but by using lodash get in place of optional chaining, the branch is now considered covered by the jest coverage report. its not a fix, just a workaround to hit 100% branch coverage per the report.

**Comment by juri-sinitson** [comment]
Instead of `let first = a?.c` I think `let first = a?.c ?? ''` is a better one for both the reader and the Code Coverage Report.
With that I see in the report, if the case `?? ''` was really covered. And then I would try to provoke the case `?? ''`. If there is no way to provoke (not even theoretically), then I would simply write `let first = a.c`.

So IMHO this issue can be closed.

**Comment by joan-saum** [comment]
@juri-sinitson , suggesting to add the operator '??' every time we want to use the optional chaining, only to have good coverage is clearly not a good solution. You don't complexify your product code just to satisfy your tests.

Same answer for you @japamatSynapse . Using a third party library only to hit the 100% branch coverage ? Yikes.

The only solution I see for now is to see when you use the optional chaining and add test for defined and undefined value, without take care of the code coverage. Code coverage does not define whether your code is well tested or not. It only defines the % of code executed by your tests. It is a nice metric to have, but definitely not a goal to reach 100%

**Comment by petarzarkov** [comment]
> Instead of `let first = a?.c` I think `let first = a?.c ?? ''` is a better one for both the reader and the Code Coverage Report.
> With that I see in the report, if the case `?? ''` was really covered. And then I would try to provoke the case `?? ''`. If there is no way to provoke (not even theoretically), then I would simply write `let first = a.c`.
> 
> So IMHO this issue can be closed.

Defaulting your value has got nothing to do with the issue here nor does it make the code "more readable".

**Comment by lcpichette** [comment]
> this worked for me if https://stackoverflow.com/a/66949633

Sadly, this fix doesn't work for my project. Any update on a fix for this? 

If, for example, `get test() {return someObject?.attribute}` is in question, and branch `?.` isn't being covered, then in theory:

```ts
it('...', () => {
    component.someObject = null;
    expect(component.test).toEqual(null);
});

it('...', () => {
    component.someObject = {attribute: 'value'};
    expect(component.test).toEqual('value');
});
```

The above should, in theory, fully cover that branch. We run through the code when `someObject` is null, and we run through the code when `someObject` is defined.

I have also added additional paranoia tests of checking if `someObject` is undefined to see if this additional test is what passes as adequate coverage.

I'm starting to draw the conclusion that either: I have something in my project incorrectly configured to support optional chaining, or, there just isn't functional support for testing coverage of this operator. 

Any updates/help on this would be awesome : )

**Comment by qxdmmc** [comment]
try the upgrade npm version.

**Comment by SaffaMeer** [comment]
what will be the recomended node version for optional chaining branch coverage?

**Comment by SaffaMeer** [comment]
i am having the same issue. branch coverage is 75% due to optional chaining,

**Comment by dimitriy-k** [comment]
any news?

**Comment by mheob** [comment]
See also duplicate #526. This issue also addresses the same problem with the nullish coalescing operator (??).

**Comment by nathanagez** [comment]
Hi there, it worked for me:

- Upgrade your node version to at least Node 14
- Update your tsconfig.js `target` attribute to `es2020`

**Comment by jiangzongyu** [comment]
any news?

**Comment by BatemanVO** [comment]
Starting on Nov. 16 with [this post](https://github.com/istanbuljs/istanbuljs/issues/516#issuecomment-971265354) in this thread, the topic was actually subtly changed - the original issue was that optional chaining is _not_ creating branches that need to be covered and is still a problem. From the original post by @chrismiceli:

> Steps to reproduce the behavior:
> Create a simple unit of code to test:
```
export const cGetter = (a) => {
  let first = a?.c;
  return first;
}
```
> And create a test that only tests the defined case:
```
import {cGetter} from './getter';
test('gets c', () => {
  expect(cGetter({c: 'hi'})).toEqual('hi');
});
```
> Expected behavior

> I would expect 50% branch coverage since the case wehre a is falsy is unchecked.

That is, a?.c should have two tests, one where `a` is defined and one where `a` is not, in order to get 100% coverage, but at the moment, only one test is required because the coverage tool doesn't consider optional chaining a branch.

However, the post on Nov. 16 switched to discussion about optional chaining behaving that way and figuring out how to get coverage on it - but what my team is seeing is that optional chaining is still not considered a branch and is allowing the team to write untested code without the tool catching it.

```
return obj.a?.b?.c?.d?.e; // 100% coverage with 1 test because none of these are considered branches
```

```
return obj.a && obj.a.b && obj.a.b.c && obj.a.b.c.d && obj.a.b.c.d.e; // Need tests for defined and undefined at each step
```

Running Node v18.9.0 as of the time of this post.

**Comment by Igorx8** [comment]
I was with the same issue, i've only changed target from ```"target": "ES6"```  to ```"target": "ES2020"``` on my tsconfig.json and it solved

**Comment by stereobooster** [comment]
Maybe related:

- https://github.com/istanbuljs/istanbuljs/issues/723
- https://github.com/istanbuljs/istanbuljs/issues/526

Different examples of how branches are calculated https://github.com/stereobooster/test-coverage-calculation

**Comment by dLo999** [comment]
Draft PR created: https://github.com/istanbuljs/istanbuljs/pull/841

This adds branch coverage instrumentation for optional chaining (`?.`) operators in `istanbul-lib-instrument`. Each `?.` now creates a binary `opt-chain` branch with nullish (short-circuit) and non-nullish (continue) paths.

Changes:
- `packages/istanbul-lib-instrument/src/visitor.js` — `OptionalMemberExpression` + `OptionalCallExpression` handlers
- `packages/istanbul-lib-instrument/test/specs/optional-chaining.yaml` — 13 new test cases

All 382 monorepo tests pass, ESLint clean.
